### PR TITLE
Add file saving pattern

### PR DESCRIPTION
**Pull request by theKashey**
Related to #42 
Adding ability to customize saving file pattern, but it is not "feature complete" as long fileName is not passed here.
Could you help me to make this real?

### CODE REVIEW

File: src/core/constants.ts [hunk 1]
```diff
@@ -31,4 +31,7 @@ export const defaultScreenshotOptions: ScreenshotOptions = {
     isLandscape: false,
   },
   knobs: {},
+  filePattern: null,
+  imageComparison: null,
+  imageThreshold: 0.2
```
Comment (tsuyoshiwada) [inline src/core/constants.ts]: What are `hoge` and `fuga` variables for?

File: src/core/utils.ts [hunk 2]
```diff
@@ -74,15 +76,27 @@ export interface Story2FilenameParams {
...
+    kind: params.kind,
+    story: params.story,
+    knobs,
+    ns,
+    vp
+  };
+  
+  const filename =
+    (params.filePattern || DEFAULT_FILE_PATTERN)
+     .replace(/\{(\w+)\}/g, (_,key) => replacements[key]);
```
Comment (tsuyoshiwada) [inline src/core/utils.ts]: I would like a unit test for this change if possible :pray:

**Comment by tsuyoshiwada** [comment]
> Could you help me to make this real?

of course :+1:

File: src/core/constants.ts (hunk 1 above)
Comment (theKashey) [reply src/core/constants.ts]: This is a part of unmade changes yet. "Usually" after the screenshooting you have to compare old and a new one(there is a lot of random text anti-alising related stuff), using [pixelmatch](https://github.com/mapbox/pixelmatch) for example. And it accepts imageThreshold.

File: src/core/utils.ts (hunk 2 above)
Comment (theKashey) [reply src/core/utils.ts]: One already exists - for default pattern. More to come.

File: src/core/constants.ts (hunk 1 above)
Comment (tsuyoshiwada) [reply src/core/constants.ts]: `storybook-chrome-screenshot` basically wants to concentrate on taking screenshots. (For wider scope of use)

There is no target for adding specific functions yet, so this time I am happy to remove this variable :pray:

Comment (theKashey) [reply src/core/constants.ts]: That could be a problem. We are using the same approach with puppeteer to capture screenshots, just not from a storybooks, and VERY often it generates _a bit_ different screenshots. And, imo, it'is this tool duty to detect AA releated changes in rendering and _hide_ them from git, for example. 
It is just a 4 lines of code to handle it.

Comment (tsuyoshiwada) [reply src/core/constants.ts]: Indeed, I understood.

If we do not add the comparative function with this PR, we would like to divide it into another PR if possible :hand:

**Comment by theKashey** [comment]
this is the last part I need. Not sure that I need exact THIS, so let me describe what I actually need:
1. Have a screenshot next to storybook. So "component" is JS + CSS + Strorybook + Static Image.
2. Not always update that screenshot, but only the changes hit's some threshold.

Actually, to fulfil both tasks it might be better to store all the images in one dirrectory, and then create some json file with information which file was generated from which story, thus enable an user to perform __any__ operation on generated screenshots.

File: src/core/app/App.ts [hunk 3]
```diff
@@ -117,6 +117,8 @@ export default class App {
       }
     ));
 
+    fs.writeFileSync(path.join(this.options.outputDir, 'stories.json'), JSON.stringify(stories));
```
Comment (theKashey) [inline src/core/app/App.ts]: This is the change I need - save stories.json, to let any other tool handle the next step.

**Comment by theKashey** [comment]
Hey @tsuyoshiwada!
I've found that I needed, and updated PR.

Now
 - one have a control on file pattern - it always useful. (and was here from the beginning)
 - result of generation is stored in "stories.json"

In case of storybook v4 that fill will look like
```json
[
  {
    "kind": "Pages/GeoAnalysis",
    "story": "Page",
    "parameters": {
      "fileName": "./src/pages/geo-analysis/GeoAnalysis.story.tsx"
    },
    "viewport": {
      "width": 1024,
      "height": 768,
      "deviceScaleFactor": 1,
      "isMobile": false,
      "hasTouch": false,
      "isLandscape": false
    },
    "knobs": {},
    "skipped": false,
    "filename": "PagesGeoAnalysis-Page.png"
  }
]
```

Next - any tool could read stories, and do whatever that tool need - for example use `filename` to move image to the target directory.
Also storybook has an API to store some additional data in story _parameters_, so - I am exposing that variable as-is.

File: src/core/utils.ts (hunk 2 above)
Comment (theKashey) [reply src/core/utils.ts]: Done! And please double check it.

File: src/core/utils.ts [hunk 4]
```diff
@@ -74,13 +76,28 @@ export interface Story2FilenameParams {
...
+    ns,
+    vp
+  };
+
+  const filename =
+    (params.filePattern || DEFAULT_FILE_PATTERN)
+      .replace(/\{(\w+)\}/g, (match, key) => filenamify(replacements[key]))
+      .replace(/-_/g, '')
+      .replace(/__/g, '')
+      .replace(/\/_/g, '');
```
Comment (theKashey) [inline src/core/utils.ts]: This moment is neat.. - I've moved the "_" and "-" you had before in values, into the file pattern,
and then I am wiping the empty placeholders from it.

File: src/models/story.ts [hunk 5]
```diff
@@ -6,15 +6,22 @@ export interface Story {
...
+export interface Parameters {
+  [key: string]: string;
+}
+
 export interface StoryWithOptions extends Story {
   namespace: string | null;
   viewport: Viewport | Viewport[];
   knobs: Knobs;
+  parameters: Parameters;
+  filePattern: string;
```
Comment (tsuyoshiwada) [inline src/models/story.ts]: Could you please add documentation for `filePattern` :smiley: ?

**Comment by theKashey** [comment]
Documentation for file pattern done.
Let me implement this for a few project, and then I'll come back and add some use cases for stories.json, and image comparison with `pixelmatch` and friends.

**Comment by tsuyoshiwada** [comment]
@theKashey Thanks! Released as `v1.2.0` :tada:

**Comment by theKashey** [comment]
Took just a two months 🚀

**Pull request merged**